src/app.js

```diff
@@ -1,104 +1,140 @@
+            if (
+              typeof (cb = cb({
+                name: path.join("."),
+                data: data
+              })) === "function"
+            ) {
+              afterActionCallbacks.push(cb)
+            }
+          })
+
+          var result = action(
+            getPath(lastPath, appState),
+            getPath(lastPath, appActions),
+            data
+          )
+
+          afterActionCallbacks.map(function(cb) {
+            result = cb(result)
+          })
 
           return typeof result === "function" ? result(update) : update(result)
         }
       } else {
-        initialize(actions[key] || (actions[key] = {}), action, name)
+        initialize(actions[key] || (actions[key] = {}), action, path)
       }
     })
+
+    function update(withState) {
+      var partialState = getPath(lastPath, appState)
+
+      if (typeof withState === "function") {
+        return update(withState(partialState))
+      }
+      if (
+        withState &&
+        (withState = setPath(
+          lastPath,
+          merge(partialState, withState),
+          appState
+        ))
+      ) {
+        requestRender((appState = withState))
+      }
+
+      return appState
+    }
+  }
+
+  function requestRender() {
+    if (appView && !renderLock) {
+      requestAnimationFrame(render, (renderLock = !renderLock))
+    }
   }
 
   function render(cb) {
     element = patch(
       appRoot,
       element,
-      oldNode,
-      (oldNode = emit("render", appView)(appState, appActions)),
+      node,
+      (node = appView(appState, appActions)),
       (renderLock = !renderLock)
     )
     while ((cb = globalInvokeLaterStack.pop())) cb()
   }
 
-  function requestRender() {
-    if (appView && !renderLock) {
-      requestAnimationFrame(render, (renderLock = !renderLock))
-    }
+  function setPath(path, value, source) {
+    var name = path[0]
+    return path.length === 0
+      ? value
+      : set(
+          name,
+          path.length > 1
+            ? setPath(
+                path.slice(1),
+                value,
+                source != null && name in source ? source[name] : {}
+              )
+            : value,
+          source
+        )
   }
 
-  function update(withState) {
-    if (typeof withState === "function") {
-      return update(withState(appState))
-    }
-    if (withState && (withState = emit("update", merge(appState, withState)))) {
-      requestRender((appState = withState))
-    }
-    return appState
+  function set(prop, value, source) {
+    var target = merge(source)
+    target[prop] = value
+    return target
   }
 
-  function emit(name, data) {
-    return (
-      (appEvents[name] || []).map(function(cb) {
-        var result = cb(appState, appActions, data)
-        if (result != null) {
-          data = result
-        }
-      }),
-      data
-    )
+  function getPath(path, source) {
+    return path.length === 0 ? source : getPath(path.slice(1), source[path[0]])
```

Comment:
@JorgeBucaran I think there's a bug here, or at least an inconsistency. Based on this code in setPath above (line 109):

```
source != null && name in source ? source[name] : {}
```

I would expect an update for a state fragment that doesn't exist yet to work, however this line will throw an error like cannot read property 'foo' of undefined from getPath. One possible fix would be to write:
```
return path.length === 0 ? source : getPath(path.slice(1), source[path[0]] || {})
```

instead. I came across the issue because the condition in setPath is currently not covered by the tests and I was trying to add coverage.